src/content/reference/rules/components-and-hooks-must-be-pure.md

```diff
@@ -0,0 +1,283 @@
+
+```js {3}
+function NewsFeed({ items }) {
+  // ✅ Array.filter is idempotent: it doesn't mutate `items`
+  const filteredItems = items.filter(item => item.isDisplayed === true);
+  return (
+    <ul>
+      {filteredItems.map(item => <li key={item.id}>{item.text}</li>)}
+    </ul>
+  );
+}
+```
+
+This means that _all_ code that runs during render must also be idempotent in order for this rule to hold. For example, this line of code is not idempotent (and therefore, neither is the component):
+
+```js {2}
+function Clock() {
+  const date = new Date(); // ❌ always returns a different result!
+  return <div>{date}</div>
+}
+```
+
+`new Date()` is not idempotent as it always returns the current date and changes its result every time it's called. When you render the above component, the time displayed on the screen will stay stuck on the time that the component was rendered. Similarly, functions like `Math.random()` also aren't idempotent, because they return different results every time they're called, even when the inputs are the same.
+
+This doesn't mean you shouldn't use non-idempotent functions like `new Date()` _at all_ – you should just avoid using them during render. One quick heuristic to tell if code runs during render is to examine where it is: if it's written at the top level like in the example above, there's a good chance it runs during render.
+
+Try building a component that displays the time in real-time in our [challenge](learn/keeping-components-pure#challenges) to see if you follow this rule!
+
+## Side effects must run outside of render {/*side-effects-must-run-outside-of-render*/}
+
+[Side effects](/learn/keeping-components-pure#side-effects-unintended-consequences) should not run in render, as React can render components multiple times to create the best possible user experience.
+
+<Note>
+Side effects are a broader term than Effects. Effects specifically refer to code that's wrapped in `useEffect`, while a side effect is a general term for code that has any observable effect other than its primary result of returning a value to the caller.
+</Note>
+
+While render must be kept pure, side effects are necessary at some point in order for your app to do anything interesting, like showing something on the screen! The key point of this rule is that side effects should not run in render, as React can render components multiple times. In most cases, you'll use [event handlers](learn/responding-to-events) to handle side effects.
+
+For example, you might have an event handler that displays a confirmation dialog after the user clicks a button. Using an event handler explicitly tells React that this code doesn't need to run during render, keeping render pure. If you've exhausted all options – and only as a last resort – you can also handle side effects using `useEffect`.
```

Comment:
possible to add code snippet here as well? 

I feel it would be better to come up with something like [ESLint's docs](https://eslint.org/docs/latest/rules/constructor-super#rule-details) with a correct and incorrect example to show which is better than the other one.

Reply:
There's already a lot of documentation around `useEffect` and mostly to not use them, so I don't want to add a new one here which might give someone the wrong impression that it's not a bad thing to use them. They're meant to be a last resort.